workspace-manager-clienttests/src/main/java/scripts/utils/WorkspaceManagerServiceUtils.java

```diff
@@ -0,0 +1,92 @@
+        }
+
+        // refresh the client service account token
+        GoogleCredentials serviceAccountCredential =
+                AuthenticationUtils.getServiceAccountCredential(
+                        server.testRunnerServiceAccount, AuthenticationUtils.userLoginScopes);
+        AccessToken accessToken = AuthenticationUtils.getAccessToken(serviceAccountCredential);
+        logger.debug(
+                "Generated access token for workspace manager service client SA: {}",
+                server.testRunnerServiceAccount.name);
+
+        // build the client object
+        ApiClient apiClient = new ApiClient();
+        apiClient.setBasePath(server.workspaceManagerUri);
+        apiClient.setAccessToken(accessToken.getTokenValue());
+
+        return apiClient;
+    }
+
+    /**
+     * Build the Workspace Manager API client object for the given test user and server specifications. This
+     * class maintains a cache of API client objects, and will return the cached object if it already
+     * exists. The token is always refreshed, regardless of whether the API client object was found in
+     * the cache or not.
+     *
+     * @param testUser the test user whose credentials are supplied to the API client object
+     * @param server the server we are testing against
+     * @return the API client object for this user
+     */
+    public static ApiClient getClientForTestUser(
+            TestUserSpecification testUser, ServerSpecification server) throws IOException {
+        if (server.workspaceManagerUri == null || server.workspaceManagerUri.isEmpty()) {
+            throw new IllegalArgumentException("Workspace Manager URI cannot be empty");
+        }
+
+        // if no test user is specified, then return a client object without an access token set
+        // this is useful if the caller wants to make ONLY unauthenticated calls
+        if (testUser == null) {
+            ApiClient apiClient = new ApiClient();
+            apiClient.setBasePath(server.workspaceManagerUri);
+        }
+
+        // refresh the user token
+        GoogleCredentials userCredential =
+                AuthenticationUtils.getDelegatedUserCredential(
+                        testUser, AuthenticationUtils.userLoginScopes);
+        AccessToken userAccessToken = AuthenticationUtils.getAccessToken(userCredential);
+
+        // TODO: have ApiClients share an HTTP client, or one per each is ok?
+        // no cached ApiClient found, so build a new one here and add it to the cache before returning
+        logger.debug(
+                "Fetching credentials and building Workspace Manager ApiClient object for test user: {}",
+                testUser.name);
+        ApiClient apiClient = new ApiClient();
+        apiClient.setBasePath(server.workspaceManagerUri);
+
+        apiClient.setAccessToken(userAccessToken.getTokenValue());
+
+        return apiClient;
+    }
```

Comment:
or, refactoring both methods (including edits to javadoc/comments):
```
   private static ApiClient buildClient(AccessToken accessToken,  ServerSpecification server) throws IOException {
        if (Strings.isNullOrEmpty(server.workspaceManagerUri)) {
            throw new IllegalArgumentException("Workspace Manager Service URI cannot be empty");
        }

        // build the client object
        ApiClient apiClient = new ApiClient();
        apiClient.setBasePath(server.workspaceManagerUri);

        if (accessToken != null) {
            apiClient.setAccessToken(accessToken.getTokenValue());
        }

        return apiClient;
    }

    /**
     * Build the Workspace Manager Service API client object for the given server specification,
     * using an access token for the test runner service account.
     *
     * @param server the server we are testing against
     * @return the API client object
     */
    public static ApiClient getClient(ServerSpecification server) throws IOException {
        if (server.testRunnerServiceAccount == null) {
            throw new IllegalArgumentException("Workspace Manager Service client service account is required");
        }

        // refresh the client service account token
        GoogleCredentials serviceAccountCredential =
                AuthenticationUtils.getServiceAccountCredential(
                        server.testRunnerServiceAccount, AuthenticationUtils.userLoginScopes);
        AccessToken accessToken = AuthenticationUtils.getAccessToken(serviceAccountCredential);
        logger.debug(
                "Generated access token for workspace manager service client SA: {}",
                server.testRunnerServiceAccount.name);

        return buildClient(accessToken, server);

    }

    /**
     * Build the Workspace Manager API client object for the given test user and server specifications.
     * The test user's token is always refreshed
     *
     * @param testUser the test user whose credentials are supplied to the API client object
     * @param server the server we are testing against
     * @return the API client object for this user
     */
    public static ApiClient getClientForTestUser(
            TestUserSpecification testUser, ServerSpecification server) throws IOException {

        AccessToken accessToken = null;

        // if no test user is specified, then return a client object without an access token set
        // this is useful if the caller wants to make ONLY unauthenticated calls
        if (testUser != null) {
            logger.debug(
                    "Fetching credentials and building Workspace Manager ApiClient object for test user: {}",
                    testUser.name);

            // refresh the user token
            GoogleCredentials userCredential =
                    AuthenticationUtils.getDelegatedUserCredential(
                            testUser, AuthenticationUtils.userLoginScopes);
            accessToken = AuthenticationUtils.getAccessToken(userCredential);
        }

        return buildClient(accessToken, server);
    }
```

Reply:
Refactored build client code.